test/test_native_mha.py

```diff
@@ -0,0 +1,301 @@
+            qs = list(torch.unbind(q))
+            if use_padding:
+                if pad_all:
+                    qs = [x[:-1] for x in qs]
+                else:
+                    qs[0] = qs[0][:-1]
+            q = torch.nested_tensor(qs, device=device, dtype=dtype)
+            if mode == "self":
+                k = v = q
+            elif mode == "encdec":
+                k = torch.nested_tensor(torch.unbind(k), device=device, dtype=dtype)
+                v = k
+            else:
+                k = torch.nested_tensor(torch.unbind(k), device=device, dtype=dtype)
+                v = torch.nested_tensor(torch.unbind(v), device=device, dtype=dtype)
+
+        ynpt, weight_npt = npt(
+            q, k, v, key_padding_mask=mask if use_padding and not use_nt else None
+        )
+        if use_nt:
+            ynpt = ynpt.to_padded_tensor(0)
+            if pad_all:
+                ynpt_final = torch.zeros_like(ypt)
+                ynpt_final[:, :ynpt.shape[1], :] = ynpt
+                ynpt = ynpt_final
+
+        def do_pad_all(tensors):
+            for t in tensors:
+                for t_i in t:
+                    t_i[-1] = torch.zeros_like(t_i[-1], device=device, dtype=dtype)
+
+        # PyTorch implementation returns non-zero junk in the padding
+        # locations; overwrite it so that the comparison works out.
+        if use_padding:
+            ypt[0][-1] = torch.zeros_like(ypt[0][-1], device=device, dtype=dtype)
+            ynpt[0][-1] = torch.zeros_like(ynpt[0][-1], device=device, dtype=dtype)
+            if pad_all:
+                do_pad_all((ypt, ynpt))
+            # Zero the last row of each TxT weight matrix
+            if need_weights:
+                if average_attn_weights:
+                    weight_pt[0][-1] = torch.zeros_like(weight_pt[0][-1], device=device, dtype=dtype)
+                    weight_npt[0][-1] = torch.zeros_like(weight_npt[0][-1], device=device, dtype=dtype)
+                    if pad_all:
+                        do_pad_all((weight_pt, weight_npt))
+                else:
+                    for nh in range(num_heads):
+                        weight_pt[0][nh][-1] = torch.zeros_like(weight_pt[0][nh][-1], device=device, dtype=dtype)
+                        weight_npt[0][nh][-1] = torch.zeros_like(weight_npt[0][nh][-1], device=device, dtype=dtype)
+
+        if dtype == torch.half:
+            torch.testing.assert_close(ypt, ynpt, atol=1e-3, rtol=1e-3)
+        else:
+            # High rtol seems necessary for
+            # test_native_multihead_attention_cpu_float32 on Windows,
+            # otherwise 2e-4 would likely be fine.
+            torch.testing.assert_close(ypt, ynpt, atol=2e-5, rtol=2e-3)
+
+        if need_weights:
+            torch.testing.assert_close(weight_pt, weight_npt)
+        else:
+            self.assertEqual(weight_pt, weight_npt)
+
+    @dtypesIfCUDA(torch.float, torch.half)
+    @dtypes(torch.float)
+    @skipMeta
+    @torch.inference_mode()
```

Comment:
no grad

Reply:
I did inference_mode on purpose. Why would we keep leaving examples of no_grad laying around when inference_mode is the better alternative people should default to?

Reply:
They are not the same though. Sometimes you want one sometimes the other.
And because inference_mode => no_grad it is better to test no_grad.

Reply:
thanks, this exposed autograd complaining about returning a tensor with use_count() > 1 in test_TransformerEncoderLayer_empty despite being run with no_grad. Not worth fighting autograd over; just cloning the input in that case instead of returning it.